File: benchmarks/js/cross_check.mjs
// Cross-checks the Rust predicate results against turf.js (ADR-0008: turf v6,
// JSTS-based, is the trusted reference; any disagreement is investigated, never
// silently accepted).
//
// Usage:
//   1. cargo build --release -p spatial-rules-benchmarks --bin cross_check
//   2. cd benchmarks/js && npm install
//   3. npm run cross-check

import { execFileSync } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { feature, booleanIntersects, booleanContains, booleanWithin, booleanOverlap } from '@turf/turf';
import booleanTouches from '@turf/boolean-touches';

const here = dirname(fileURLToPath(import.meta.url));
const pairsFile = join(here, '..', 'data', 'cross_check.json');
const fixtures = JSON.parse(readFileSync(pairsFile, 'utf8'));

const rustBin =
  process.env.CROSS_CHECK_BIN ??
  join(here, '..', '..', 'target', 'release', process.platform === 'win32' ? 'cross_check.exe' : 'cross_check');

const rust = JSON.parse(execFileSync(rustBin, [pairsFile], { encoding: 'utf8' }));

// turf v6 `booleanOverlap`/`booleanTouches` deviate from DE-9IM on these cases
// (geo/DE-9IM is authoritative per ADR-0008, verified against JTS/GEOS):
//   - `touching_edge`/`touching_corner` `.overlaps`: turf reports true for
//     boundary-touch; DE-9IM `overlaps` requires a same-dimension interior
//     intersection, so it is false.
//   - `identical` `.touches`: turf reports true for identical polygons;
//     DE-9IM `touches` requires disjoint interiors, so it is false.
const knownQuirks = new Set([
  'touching_edge.overlaps',
  'touching_corner.overlaps',
  'identical.touches',
]);

let mismatches = 0;
for (const pair of fixtures.pairs) {
  const a = feature(pair.a);
  const b = feature(pair.b);

  // turf v6 `booleanContains` cannot take a MultiPolygon/GeometryCollection as
  // the contained geometry (feature2) — a known turf limitation (ADR-0008
  // "known quirk"). `intersects`, `within`, `touches` and `overlaps` cover
  // those pairs fully; the skipped `contains` value is hand-verified against
  // DE-9IM.
  const supportsContains = !['MultiPolygon', 'GeometryCollection'].includes(pair.b.type);
  const turf = {
    intersects: booleanIntersects(a, b),
    within: booleanWithin(a, b),
    touches: booleanTouches(a, b),
    overlaps: booleanOverlap(a, b),
  };
  if (supportsContains) {
    turf.contains = booleanContains(a, b);
    // turf has no `covers`/`covered_by` boolean module. For area fixtures
    // (all of ours), DE-9IM `covers` ≡ `contains` and `covered_by` ≡ `within`,
    // so those oracles double as the covers/covered_by oracles (ADR-0012).
    turf.covers = turf.contains;
    turf.covered_by = turf.within;
  }

  const rustResult = rust.pairs.find((entry) => entry.name === pair.name);
  if (!rustResult) {
    console.error(`missing rust result for ${pair.name}`);
    process.exit(1);
  }
  const predicates = supportsContains
    ? ['intersects', 'contains', 'within', 'covers', 'covered_by', 'touches', 'overlaps']
    : ['intersects', 'within', 'touches', 'overlaps'];
  for (const predicate of predicates) {
    if (knownQuirks.has(`${pair.name}.${predicate}`)) {
      continue;
    }
    if (rustResult[predicate] !== turf[predicate]) {
      mismatches += 1;
      console.error(
        `MISMATCH ${pair.name}.${predicate}: rust=${rustResult[predicate]} turf=${turf[predicate]}`,
      );
    }
  }
  const containsText = supportsContains ? ` contains=${turf.contains}` : ' contains=skipped(turf-v6-limit)';
  console.log(
    `${pair.name}: intersects=${turf.intersects}${containsText} within=${turf.within} covers=${turf.covers ?? 'skipped'} covered_by=${turf.covered_by ?? 'skipped'} touches=${turf.touches} overlaps=${turf.overlaps}`,
  );
}

if (mismatches > 0) {
  console.error(`${mismatches} mismatch(es)`);
  process.exit(1);
}
console.log('cross-check green');

File: benchmarks/src/bin/cross_check.rs
//! Emits the engine's spatial-predicate verdicts for the turf.js cross-check
//! fixtures (ticket 12, ADR-0008, ADR-0012). Reads
//! `benchmarks/data/cross_check.json` and, for each named pair, drives the
//! engine through its public seams — a single-rule ruleset and a single
//! candidate, queried once per predicate — so the cross-check certifies the
//! engine's `SpatialPredicate → DE-9IM` mapping (architecture-hardening 08),
//! not just geo's raw matrix helpers.

use spatial_rules_core::{Candidate, Query, Rule, Ruleset, SpatialPredicate};

const PREDICATES: [SpatialPredicate; 7] = [
    SpatialPredicate::Intersects,
    SpatialPredicate::Contains,
    SpatialPredicate::Within,
    SpatialPredicate::Covers,
    SpatialPredicate::CoveredBy,
    SpatialPredicate::Touches,
    SpatialPredicate::Overlaps,
];

fn main() {
    let path = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "benchmarks/data/cross_check.json".to_string());
    let text = std::fs::read_to_string(&path).expect("read cross_check.json");
    let root: serde_json::Value = serde_json::from_str(&text).expect("parse cross_check.json");
    let pairs = root["pairs"].as_array().expect("pairs array");

    let mut results = Vec::new();
    for pair in pairs {
        let name = pair["name"].as_str().expect("pair name");
        let candidate = Candidate::new("a".to_string(), geometry_from_json(&pair["a"]));
        let rule = Rule {
            id: "b".to_string(),
            properties: Default::default(),
            geometry: geometry_from_json(&pair["b"]),
        };
        let ruleset = Ruleset::build(vec![rule]).expect("fixture rule must build");

        let mut result = serde_json::Map::new();
        result.insert(
            "name".to_string(),
            serde_json::Value::String(name.to_string()),
        );
        for predicate in PREDICATES {
            let mask = ruleset.query_mask(
                std::slice::from_ref(&candidate),
                &Query::new(predicate),
            );
            // The engine mask: 1 = matched, 0 = not matched, 2 = invalid.
            result.insert(
                predicate.as_str().to_string(),
                serde_json::Value::Bool(mask[0] == 1),
            );
        }
        results.push(serde_json::Value::Object(result));
    }

    println!("{}", serde_json::json!({ "pairs": results }));
}

fn geometry_from_json(value: &serde_json::Value) -> geo::Geometry<f64> {
    let geojson: geojson::GeoJson = serde_json::from_value(value.clone()).expect("GeoJSON");
    let geojson::GeoJson::Geometry(geometry) = geojson else {
        panic!("expected a geometry object");
    };
    geo::Geometry::try_from(&geometry).expect("convert to geo geometry")
}
